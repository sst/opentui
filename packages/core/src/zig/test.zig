const std = @import("std");

// Import all test modules
const text_buffer_tests = @import("tests/text-buffer_test.zig");
const text_buffer_highlights_tests = @import("tests/text-buffer-highlights_test.zig");
const text_buffer_view_tests = @import("tests/text-buffer-view_test.zig");
const text_buffer_selection_tests = @import("tests/text-buffer-selection_test.zig");
const text_buffer_drawing_tests = @import("tests/text-buffer-drawing_test.zig");
const text_buffer_segment_tests = @import("tests/text-buffer-segment_test.zig");
const text_buffer_iterators_tests = @import("tests/text-buffer-iterators_test.zig");
const edit_buffer_tests = @import("tests/edit-buffer_test.zig");
const edit_buffer_history_tests = @import("tests/edit-buffer-history_test.zig");
const editor_view_tests = @import("tests/editor-view_test.zig");
const grapheme_tests = @import("tests/grapheme_test.zig");
const syntax_style_tests = @import("tests/syntax-style_test.zig");
const rope_tests = @import("tests/rope_test.zig");
const rope_nested_tests = @import("tests/rope-nested_test.zig");
const rope_fuzz_tests = @import("tests/rope_fuzz_test.zig");
const utf8_tests = @import("tests/utf8_test.zig");
const event_emitter_tests = @import("tests/event-emitter_test.zig");
const buffer_tests = @import("tests/buffer_test.zig");
// const example_tests = @import("example_test.zig");

// Re-export test declarations from individual test files
// This allows `zig test index.zig` to run all tests
comptime {
    _ = text_buffer_tests;
    _ = text_buffer_highlights_tests;
    _ = text_buffer_view_tests;
    _ = text_buffer_selection_tests;
    _ = text_buffer_drawing_tests;
    _ = text_buffer_segment_tests;
    _ = text_buffer_iterators_tests;
    _ = edit_buffer_tests;
    _ = edit_buffer_history_tests;
    _ = editor_view_tests;
    _ = grapheme_tests;
    _ = syntax_style_tests;
    _ = rope_tests;
    _ = rope_nested_tests;
    _ = rope_fuzz_tests;
    _ = utf8_tests;
    _ = event_emitter_tests;
    _ = buffer_tests;
    // _ = example_tests;
}
