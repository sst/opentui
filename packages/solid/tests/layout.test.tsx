import { describe, expect, it, beforeEach, afterEach } from "bun:test"
import { testRender } from "../index"
import { createSignal, Show } from "solid-js"

let testSetup: Awaited<ReturnType<typeof testRender>>

describe("SolidJS Renderer Integration Tests", () => {
  beforeEach(async () => {
    if (testSetup) {
      testSetup.renderer.destroy()
    }
  })

  afterEach(() => {
    if (testSetup) {
      testSetup.renderer.destroy()
    }
  })

  describe("Basic Text Rendering", () => {
    it("should render simple text correctly", async () => {
      testSetup = await testRender(() => <text>Hello World</text>, {
        width: 20,
        height: 5,
      })

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should render multiline text correctly", async () => {
      testSetup = await testRender(
        () => (
          <text>
            Line 1
            <br />
            Line 2
            <br />
            Line 3
          </text>
        ),
        {
          width: 15,
          height: 5,
        },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should render text with dynamic content", async () => {
      const counter = () => 42

      testSetup = await testRender(() => <text>Counter: {counter()}</text>, {
        width: 20,
        height: 3,
      })

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })
  })

  describe("Box Layout Rendering", () => {
    it("should render basic box layout correctly", async () => {
      testSetup = await testRender(
        () => (
          <box style={{ width: 20, height: 5, border: true }}>
            <text>Inside Box</text>
          </box>
        ),
        {
          width: 25,
          height: 8,
        },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should render nested boxes correctly", async () => {
      testSetup = await testRender(
        () => (
          <box style={{ width: 30, height: 10, border: true }} title="Parent Box">
            <box style={{ left: 2, top: 2, width: 10, height: 3, border: true }}>
              <text>Nested</text>
            </box>
            <text style={{ left: 15, top: 2 }}>Sibling</text>
          </box>
        ),
        {
          width: 35,
          height: 12,
        },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should render absolute positioned boxes", async () => {
      testSetup = await testRender(
        () => (
          <>
            <box
              style={{
                position: "absolute",
                left: 0,
                top: 0,
                width: 10,
                height: 3,
                border: true,
                backgroundColor: "red",
              }}
            >
              <text>Box 1</text>
            </box>
            <box
              style={{
                position: "absolute",
                left: 12,
                top: 2,
                width: 10,
                height: 3,
                border: true,
                backgroundColor: "blue",
              }}
            >
              <text>Box 2</text>
            </box>
          </>
        ),
        {
          width: 25,
          height: 8,
        },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })
  })

  describe("Reactive Updates", () => {
    it("should handle reactive state changes", async () => {
      const [counter, setCounter] = createSignal(0)

      testSetup = await testRender(() => <text>Counter: {counter()}</text>, {
        width: 15,
        height: 3,
      })

      await testSetup.renderOnce()
      const initialFrame = testSetup.captureCharFrame()

      setCounter(5)
      await testSetup.renderOnce()
      const updatedFrame = testSetup.captureCharFrame()

      expect(initialFrame).toMatchSnapshot()
      expect(updatedFrame).toMatchSnapshot()
      expect(updatedFrame).not.toBe(initialFrame)
    })

    it("should handle conditional rendering", async () => {
      const [showText, setShowText] = createSignal(true)

      testSetup = await testRender(
        () => (
          <text>
            Always visible
            <Show when={showText()} fallback="">
              {" - Conditional text"}
            </Show>
          </text>
        ),
        {
          width: 30,
          height: 3,
        },
      )

      await testSetup.renderOnce()
      const visibleFrame = testSetup.captureCharFrame()

      setShowText(false)
      await testSetup.renderOnce()
      const hiddenFrame = testSetup.captureCharFrame()

      expect(visibleFrame).toMatchSnapshot()
      expect(hiddenFrame).toMatchSnapshot()
      expect(hiddenFrame).not.toBe(visibleFrame)
    })
  })

  describe("Complex Layouts", () => {
    it("should render complex nested layout correctly", async () => {
      testSetup = await testRender(
        () => (
          <box style={{ width: 40, border: true }} title="Complex Layout">
            <box style={{ left: 2, width: 15, height: 5, border: true, backgroundColor: "#333" }}>
              <text style={{ fg: "cyan" }}>Header Section</text>
              <text style={{ fg: "yellow" }}>Menu Item 1</text>
              <text style={{ fg: "yellow" }}>Menu Item 2</text>
            </box>
            <box style={{ left: 18, width: 18, height: 8, border: true, backgroundColor: "#222" }}>
              <text style={{ fg: "green" }}>Content Area</text>
              <text style={{ fg: "white" }}>Some content here</text>
              <text style={{ fg: "white" }}>More content</text>
              <text style={{ fg: "magenta" }}>Footer text</text>
            </box>
            <text style={{ left: 2, fg: "gray" }}>Status: Ready</text>
          </box>
        ),
        {
          width: 45,
          height: 18,
        },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should render text with mixed styling and layout", async () => {
      testSetup = await testRender(
        () => (
          <box style={{ width: 35, height: 8, border: true }}>
            <text>
              <span style={{ fg: "red", bold: true }}>ERROR:</span> Something went wrong
            </text>
            <text>
              <span style={{ fg: "yellow" }}>WARNING:</span> Check your settings
            </text>
            <text>
              <span style={{ fg: "green" }}>SUCCESS:</span> All systems operational
            </text>
          </box>
        ),
        {
          width: 40,
          height: 10,
        },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should render scrollbox with sticky scroll and spacer", async () => {
      testSetup = await testRender(
        () => (
          <box maxHeight={"100%"} maxWidth={"100%"}>
            <scrollbox
              scrollbarOptions={{ visible: false }}
              stickyScroll={true}
              stickyStart="bottom"
              paddingTop={1}
              paddingBottom={1}
              title="scroll area"
              rootOptions={{
                flexGrow: 0,
              }}
              border
            >
              <box border height={10} title="hi" />
            </scrollbox>
            <box border height={10} title="spacer">
              <text>spacer</text>
            </box>
          </box>
        ),
        {
          width: 30,
          height: 25,
        },
      )

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })
  })

  describe("Empty and Edge Cases", () => {
    it("should handle empty component", async () => {
      testSetup = await testRender(() => <></>, {
        width: 10,
        height: 5,
      })

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should handle component with no children", async () => {
      testSetup = await testRender(() => <box style={{ width: 10, height: 5 }} />, {
        width: 15,
        height: 8,
      })

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })

    it("should handle very small dimensions", async () => {
      testSetup = await testRender(() => <text>Hi</text>, {
        width: 5,
        height: 3,
      })

      await testSetup.renderOnce()
      const frame = testSetup.captureCharFrame()
      expect(frame).toMatchSnapshot()
    })
  })
})
