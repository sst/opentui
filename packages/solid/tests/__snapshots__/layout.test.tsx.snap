// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`SolidJS Renderer Integration Tests Basic Text Rendering should render simple text correctly 1`] = `
"Hello World         
                    
                    
                    
                    
"
`;

exports[`SolidJS Renderer Integration Tests Basic Text Rendering should render multiline text correctly 1`] = `
"Line 1         
Line 2         
Line 3         
               
               
"
`;

exports[`SolidJS Renderer Integration Tests Basic Text Rendering should render text with dynamic content 1`] = `
"Counter: 42         
                    
                    
"
`;

exports[`SolidJS Renderer Integration Tests Box Layout Rendering should render basic box layout correctly 1`] = `
"┌──────────────────┐     
│Inside Box        │     
│                  │     
│                  │     
└──────────────────┘     
                         
                         
                         
"
`;

exports[`SolidJS Renderer Integration Tests Box Layout Rendering should render nested boxes correctly 1`] = `
"┌─Parent Box─────────────────┐     
│                            │     
│                            │     
│  ┌────────┐                │     
│  │Nested  │                │     
│  └────────┘                │     
│               Sibling      │     
│                            │     
│                            │     
└────────────────────────────┘     
                                   
                                   
"
`;

exports[`SolidJS Renderer Integration Tests Box Layout Rendering should render absolute positioned boxes 1`] = `
"┌────────┐               
│Box 1   │               
└────────┘  ┌────────┐   
            │Box 2   │   
            └────────┘   
                         
                         
                         
"
`;

exports[`SolidJS Renderer Integration Tests Reactive Updates should handle reactive state changes 1`] = `
"Counter: 0     
               
               
"
`;

exports[`SolidJS Renderer Integration Tests Reactive Updates should handle reactive state changes 2`] = `
"Counter: 5     
               
               
"
`;

exports[`SolidJS Renderer Integration Tests Reactive Updates should handle conditional rendering 1`] = `
"Always visible - Conditional t
                              
                              
"
`;

exports[`SolidJS Renderer Integration Tests Reactive Updates should handle conditional rendering 2`] = `
"Always visible                
                              
                              
"
`;

exports[`SolidJS Renderer Integration Tests Complex Layouts should render complex nested layout correctly 1`] = `
"┌─Complex Layout───────────────────────┐     
│  ┌─────────────┐                     │     
│  │Header Sectio│                     │     
│  │Menu Item 1  │                     │     
│  │Menu Item 2  │                     │     
│  └─────────────┘                     │     
│                  ┌────────────────┐  │     
│                  │Content Area    │  │     
│                  │Some content her│  │     
│                  │More content    │  │     
│                  │Footer text     │  │     
│                  │                │  │     
│                  │                │  │     
│                  └────────────────┘  │     
│  Status: Ready                       │     
└──────────────────────────────────────┘     
                                             
                                             
"
`;

exports[`SolidJS Renderer Integration Tests Complex Layouts should render text with mixed styling and layout 1`] = `
"┌─────────────────────────────────┐     
│ERROR: Something went wrong      │     
│WARNING: Check your settings     │     
│SUCCESS: All systems operational │     
│                                 │     
│                                 │     
│                                 │     
└─────────────────────────────────┘     
                                        
                                        
"
`;

exports[`SolidJS Renderer Integration Tests Complex Layouts should render scrollbox with sticky scroll and spacer 1`] = `
"┌─scroll area────────────────┐
│                            │
│┌─hi───────────────────────┐│
││                          ││
││                          ││
││                          ││
││                          ││
││                          ││
││                          ││
││                          ││
││                          ││
│└──────────────────────────┘│
│                            │
│                            │
└────────────────────────────┘
┌─spacer─────────────────────┐
│spacer                      │
│                            │
│                            │
│                            │
│                            │
│                            │
│                            │
│                            │
└────────────────────────────┘
"
`;

exports[`SolidJS Renderer Integration Tests Empty and Edge Cases should handle empty component 1`] = `
"          
          
          
          
          
"
`;

exports[`SolidJS Renderer Integration Tests Empty and Edge Cases should handle component with no children 1`] = `
"               
               
               
               
               
               
               
               
"
`;

exports[`SolidJS Renderer Integration Tests Empty and Edge Cases should handle very small dimensions 1`] = `
"Hi   
     
     
"
`;
